src/Analyzer/Passes/LogicalExpressionOptimizerPass.cpp

```diff
@@ -590,6 +590,14 @@ class JoinOnLogicalExpressionOptimizerVisitor : public InDepthQueryTreeVisitorWi
 
     void enterImpl(QueryTreeNodePtr & node)
     {
+        /** Alias to expression in JOIN ON section may be used in SELECT.
+          * Optimization that is safe for JOIN ON may not be safe for SELECT, for example:
+          * Values `NULL` and `false` are not equivalent in SELECT, so we cannot change type from Nullable(UInt8) to UInt8 there, while it's valid for `JOIN ON`.
+          * Also, operator <=> can be used in JOIN ON, but not in SELECT, so we need to keep original expression `a = b OR isNull(a) AND isNull(b) there.
+          */
+        if (node.use_count() > 1)
```

Comment:
i just learned that we have a special helper for that :)
https://github.com/ClickHouse/ClickHouse/blob/master/base/base/isSharedPtrUnique.h

Reply:
By the way, could it be a measurable slowdown if we'd `clone` unconditionally?

Reply:
It's not a helper; it's a crutch. Ideally not to rely on this method in a multithreaded code of course.

Reply:
It turns out that the issue can be fixed with https://github.com/ClickHouse/ClickHouse/pull/66143. Once that's resolved, we can remove this change.
I'll merge the current PR for now, as it was fuzzed by tests in another PR. Later, I'll check whether we need to verify `use_count` or if we can simply remove `clone` entirely.